README.md
# NextRJ Utilities

The utilities for `path`, `string`, `duration`.

## Usage

### Object functions

```ts
import { recursiveAssign } from "https://deno.land/x/nextrj_utils@$VERSION/object.js"
const obj1 = { id: 1, name: "test1", nested: { num: 1, keep: true } }
const obj2 = { name: "test2", nested: { num: 2, ext: "more" } }
const obj3 = recursiveAssign({}, obj1, obj2)
console.log(obj3) // { id: 1, name: "test2", nested: { num: 2, keep: true, ext: "more" } }
```

### JsonP functions

```ts
import { parse } from "https://deno.land/x/nextrj_utils@$VERSION/jsonp.ts"
const j = parse("jsonpCallbackName({ok: true})") as Record<string, boolean>
assertStrictEquals(j.ok, true)
```

### Path functions

```ts
import { exists, existsSync } from "https://deno.land/x/nextrj_utils@$VERSION/path.ts"
const pathExists = await exists("./foo/bar")
const pathExistsSync = existsSync("./foo/bar")
console.log(pathExists) // true or false
console.log(pathExistsSync) // true or false
```

### String functions

**Format string template:**

```ts
import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
import { format } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"

assertStrictEquals(format("${a} / ${b}", { a: 50, b: 100 }), "50 / 100")

const f = (v: number): number => ++v
assertStrictEquals(format("${v}-${f(1)}", { v: 1, f }), "1-2")
```

**Truncate filename:**

```ts
import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
import { truncateFilename } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"

truncateFilename("123456789.zip", 10) // "12...9.zip"
// 🦄 is 2 column width
truncateFilename("1234567🦄.zip", 10) // "1...🦄.zip"
```

**Truncate string:**

```ts
import { truncate } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"

const maxLen = 10
// defaults truncate by max-column-count for better match visable with.
console.log(truncate("=".repeat(1000), maxLen))
console.log(truncate("±".repeat(1000), maxLen))
console.log(truncate("★".repeat(1000), maxLen))
console.log(truncate("中".repeat(1000), maxLen))
console.log(truncate("🦄".repeat(1000), maxLen))
console.log(truncate("🦄".repeat(1000), maxLen))
console.log(truncate("=中±🦄±中=".repeat(1000), maxLen))

// they should all have the very similar visable width in the terminal:
```

![](./assets/string_truncate_01.png)

### Format seconds to duration

```ts
import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
import { format } from "https://deno.land/x/nextrj_utils@$VERSION/duration.ts"

// mm:ss
assertStrictEquals(format(15), "00:15")
assertStrictEquals(format(2 * 60 + 15), "02:15")

// hh:mm:ss
assertStrictEquals(format(13 * 60 * 60 + 2 * 60 + 15), "13:02:15")
assertStrictEquals(format(24 * 60 * 60 + 2 * 60 + 15), "24:02:15")
```

### Fetch remote file with progress callback

```ts
import { Fetcher } from "https://deno.land/x/nextrj_utils@$VERSION/file.ts"

const fetcher = new Fetcher(
  "https://www.example.com/test.pdf",
  {
    to: "example-test.pdf",
    on: {
      start(total: number) {
        console.log(`start: total=${total}`)
      },
      received(received: number, total: number) {
        console.log(`received: received=${received}, total=${total}`)
      },
      end(success: boolean) {
        console.log(`end: success=${success}`)
      },
    },
  },
)
await fetcher.fetch()
```


CHANGELOG.md
# NextRJ Utilities Changelog

## 2023-12-19 0.12.0

- Upgrade to `deno.land/std@0.209.0`

## 2023-04-14 0.11.0

- Add method `object.js/recursiveAssign` for recursive assign object property value.
  ```ts
  import { recursiveAssign } from "https://deno.land/x/nextrj_utils@$VERSION/object.js"
  const obj1 = { id: 1, name: "test1", nested: { num: 1, keep: true } }
  const obj2 = { name: "test2", nested: { num: 2, ext: "more" } }
  const obj3 = recursiveAssign({}, obj1, obj2)
  //-> { id: 1, name: "test2", nested: { num: 2, keep: true, ext: "more" } }
  ```

## 2023-04-12 0.10.0

- Add method `jsonp.ts/parse` for get jsonp response value
  ```ts
  parse("callback({ok: true})") // {ok: true}
  parse("parse("callback([1, 2])")") // [1, 2]
  ```

## 2023-03-21 0.9.0

- Add config property `toIsNotTemplate` to `FetcherInit`.

## 2023-03-20 0.8.0

- Fixed `file.ts/getLastPathName(url)` error again.

## 2023-03-20 0.7.0

- Fixed `file.ts/getLastPathName(url)` auto encode url error.

## 2023-03-19 0.6.4

- Change `npm:content-disposition` to `https://cdn.skypack.dev/content-disposition`
  > Because `deno compile` not support npm specifiers yet.

## 2023-03-16 0.6.3

- Consumed fetch response body when not ok.

## 2023-03-16 0.6.2

- Fixed truncateFilename always has `...`

## 2023-03-16 0.6.1

- Enhance FetcherInit.on.start to has fileName and filePath params.

## 2023-03-16 0.6.0

- Upgrade to `deno/std@0.180.0`.

## 2023-03-16 0.5.0

- Enhance getLastPathName.
- Add getFileNameFromContentDisposition method.
- Add getFileName method.
- Make `Fetcher.fetch` return file path and `options.to` support template.

## 2023-03-05 0.4.0

- Add file fetcher with progress callback (start, received, end).

## 2023-02-26 0.3.1

- Polishing document.

## 2023-02-26 0.3.0

- Add method `string.ts/truncateFileName` for shorten a filename with its extension
  ```ts
  truncateFilename("123456789.zip", 10) // "12...9.zip"
  // 🦄 is 2 column width
  truncateFilename("1234567🦄.zip", 10) // "1...🦄.zip"
  ```

## 2023-02-25 0.2.1

- Fixed `string.ts` document error.

## 2023-02-25 0.2.0

- Rename function `codePointLength` to `wordCount`.
- Add method `string.ts/columnCount` for get the visable width of a string.
- Refactor method `string.ts/truncate` to default truncate by max-column-count.

## 2023-02-22 0.1.0

- Format seconds to duration (`duration.ts`).
  - assertStrictEquals(format(2 * 60 + 15), "02:15")
  - assertStrictEquals(format(13 * 60 * 60 + 2 * 60 + 15), "13:02:15")

## 2023-02-19 0.0.5

- Add string template format function.
  > `format("${a}/${b}", {a: 50, b: 100}) // "50/100"`
- Polishing README
- Fixed `path.ts` document error

## 2023-02-05 0.0.4

- Fixed `string.ts` document error again.

## 2023-02-05 0.0.3

- Fixed `string.ts` document error.

## 2023-02-05 0.0.2

- Add method `string.ts/truncate` for truncate string by max-byte-length or max-code-point-length.
- Add method `string.ts/codePointLength` for get the string's code-point-length.
  > This is not the same code-unit-length of `String.length` for multi-byte string.
- Add method `string.ts/byteLength` for get the string's byte-length.

## 2023-02-05 0.0.1

- Add method `path.ts/exists|existsSync` for check whether the path is exists.


file_test.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.
import { assertRejects, assertStrictEquals, joinPath } from "./deps.ts"
import { existsSync } from "./path.ts"
import { Fetcher, getLastPathName } from "./file.ts"

Deno.test("get last path name", () => {
  // url
  assertStrictEquals(getLastPathName("https://test.com/xxx?a=b"), "xxx")
  assertStrictEquals(getLastPathName("https://test.com/xxx"), "xxx")
  assertStrictEquals(getLastPathName("https://test.com/xxx/"), "xxx")
  assertStrictEquals(getLastPathName("https://test.com/xxx/?a=b"), "xxx")
  assertStrictEquals(getLastPathName("https://test.com/"), "test.com")
  assertStrictEquals(getLastPathName("https://test.com/?a=b"), "test.com")
  assertStrictEquals(getLastPathName("https://test.com"), "test.com")
  assertStrictEquals(getLastPathName("https://test.com?a=b"), "test.com")
  assertStrictEquals(getLastPathName("https://test.com/中文"), "中文")
  assertStrictEquals(getLastPathName("file://test.com/xxx?a=b"), "xxx")
  assertStrictEquals(getLastPathName("file://test.com/xxx"), "xxx")
  assertStrictEquals(getLastPathName("file://test.com/xxx/"), "xxx")
  assertStrictEquals(getLastPathName("file://test.com/xxx/?a=b"), "xxx")
  assertStrictEquals(getLastPathName("file://test.com/中文"), "中文")
  assertStrictEquals(getLastPathName("file://test.com/"), "test.com")
  assertStrictEquals(getLastPathName("file://test.com"), "test.com")
  assertStrictEquals(getLastPathName("file://D:/xxx"), "xxx")
  assertStrictEquals(getLastPathName("file://D:\\xxx"), "xxx")

  // path
  assertStrictEquals(getLastPathName("/test.com/xxx.y"), "xxx.y")
  assertStrictEquals(getLastPathName("/test.com/xxx.y/"), "xxx.y")
  assertStrictEquals(getLastPathName("test.com/xxx.y"), "xxx.y")
  assertStrictEquals(getLastPathName("test.com/xxx.y/"), "xxx.y")
  assertStrictEquals(getLastPathName("abc"), "abc")
  assertStrictEquals(getLastPathName("中文"), "中文")
  assertStrictEquals(getLastPathName("test.com/中文"), "中文")
  assertStrictEquals(getLastPathName("D:/test.com/中文"), "中文")
})

Deno.test("fetch to default filename", async () => {
  const from = "assets/string_truncate_01.png"
  const to = "temp/string_truncate_01.png" // fileName extract from url
  const total = Deno.statSync(from).size

  // clean
  if (existsSync(to)) Deno.removeSync(to)

  // fetch
  let callStart = 0, callReceived = 0, callEnd = 0
  const fetcher = new Fetcher(
    `file://${Deno.realPathSync(from)}`,
    {
      on: {
        start(total: number) {
          //console.log(`start: total=${total}`)
          callStart++
          assertStrictEquals(total, -1)
        },
        received(received: number, total: number) {
          //console.log(`received: received=${received}, total=${total}`)
          callReceived = received
          assertStrictEquals(total, -1)
          assertStrictEquals(received > 0, true)
        },
        end(success: boolean) {
          //console.log(`end: success=${success}`)
          callEnd++
          assertStrictEquals(success, true)
        },
      },
    },
  )
  const filePath = await fetcher.fetch()

  // verify
  assertStrictEquals(filePath, to)
  assertStrictEquals(existsSync(filePath), true)
  assertStrictEquals(Deno.statSync(filePath).size, total)
  assertStrictEquals(callStart, 1)
  assertStrictEquals(callEnd, 1)
  assertStrictEquals(callReceived, total)

  // clean
  Deno.removeSync(filePath)
})

Deno.test("fetch to specific filename", async () => {
  const from = "file.ts"
  const to = joinPath("temp", from).replace(/\\/g, "/")
  const total = Deno.statSync(from).size

  // clean
  if (existsSync(to)) Deno.removeSync(to)

  // fetch
  let callStart = 0, callReceived = 0, callEnd = 0
  const fetcher = new Fetcher(
    `file://${Deno.realPathSync(from)}`,
    {
      to,
      on: {
        start(total: number) {
          callStart++
          assertStrictEquals(total, -1)
        },
        received(received: number, total: number) {
          callReceived = received
          assertStrictEquals(total, -1)
          assertStrictEquals(received > 0, true)
        },
        end(success: boolean) {
          callEnd++
          assertStrictEquals(success, true)
        },
      },
    },
  )
  const filePath = await fetcher.fetch()

  // verify
  assertStrictEquals(filePath, to)
  assertStrictEquals(existsSync(to), true)
  assertStrictEquals(Deno.statSync(to).size, total)
  assertStrictEquals(callStart, 1)
  assertStrictEquals(callEnd, 1)
  assertStrictEquals(callReceived, total)

  // clean
  Deno.removeSync(to)
})

Deno.test("fetch remote file", async () => {
  const from = "https://docs.spring.io/spring-framework/docs/6.0.6/reference/pdf/spring-framework.pdf"
  const to = joinPath("temp", "spring-framework.pdf").replace(/\\/g, "/")
  const total = 20814747

  // clean
  if (existsSync(to)) Deno.removeSync(to)

  // fetch
  let callStart = 0, callReceived = 0, callEnd = 0
  const fetcher = new Fetcher(
    from,
    {
      to,
      on: {
        start(total: number) {
          callStart++
          assertStrictEquals(total, total)
        },
        received(received: number, total: number) {
          callReceived = received
          assertStrictEquals(total, total)
          assertStrictEquals(received > 0, true)
        },
        end(success: boolean) {
          callEnd++
          assertStrictEquals(success, true)
        },
      },
    },
  )
  const filePath = await fetcher.fetch()

  // verify
  assertStrictEquals(filePath, to)
  assertStrictEquals(existsSync(to), true)
  assertStrictEquals(Deno.statSync(to).size, total)
  assertStrictEquals(callStart, 1)
  assertStrictEquals(callEnd, 1)
  assertStrictEquals(callReceived, total)

  // clean
  Deno.removeSync(to)
})

Deno.test("fetch file with template options.to", async () => {
  const from = "file.ts"
  const to = "temp/file.ts"
  const total = Deno.statSync(from).size

  // clean
  if (existsSync(to)) Deno.removeSync(to)

  // fetch
  let callStart = 0, callReceived = 0, callEnd = 0
  const fetcher = new Fetcher(
    `file://${Deno.realPathSync(from)}`,
    {
      to: "temp/${fileName}",
      on: {
        start(total: number, fileName: string, filePath: string) {
          console.log(`fileName=${fileName}, filePath=${filePath}`)
          callStart++
          assertStrictEquals(total, total)
        },
        received(received: number, total: number) {
          callReceived = received
          assertStrictEquals(total, total)
          assertStrictEquals(received > 0, true)
        },
        end(success: boolean) {
          callEnd++
          assertStrictEquals(success, true)
        },
      },
    },
  )
  const filePath = await fetcher.fetch()

  // verify
  assertStrictEquals(filePath, to)
  assertStrictEquals(existsSync(to), true)
  assertStrictEquals(Deno.statSync(to).size, total)
  assertStrictEquals(callStart, 1)
  assertStrictEquals(callEnd, 1)
  assertStrictEquals(callReceived, total)

  // clean
  Deno.removeSync(to)
})

Deno.test("fetch error by unknown domain", async () => {
  await assertRejects(
    () => new Fetcher("https://unknown.unknown/").fetch(),
    TypeError,
    "error sending request for url",
  )
})

Deno.test("fetch response 403 Forbidden", async () => {
  await assertRejects(
    () => new Fetcher("https://v26-web.douyinvod.com").fetch(),
    Error,
    "403 Forbidden",
  )
})


jsonp_test.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.
import { assertStrictEquals } from "./deps.ts"
import { parse } from "./jsonp.ts"

Deno.test("parse to string", () => {
  assertStrictEquals(parse('callback("abc")'), "abc")
})

Deno.test("parse to boolean", () => {
  assertStrictEquals(parse("callback(true)"), true)
  assertStrictEquals(parse("callback(false)"), false)
})

Deno.test("parse to number", () => {
  assertStrictEquals(parse("callback(2)"), 2)
  assertStrictEquals(parse("callback(1.01)"), 1.01)
})

Deno.test("parse to jsonArray", () => {
  const array = parse("callback([1, 2])") as Array<number>
  assertStrictEquals(array[0], 1)
  assertStrictEquals(array[1], 2)
  assertStrictEquals(array.length, 2)
})

Deno.test("parse to jsonObject", () => {
  const j = parse("callback({ok: true})") as Record<string, boolean>
  assertStrictEquals(j.ok, true)
})


object_test.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.
import { assertStrictEquals } from "./deps.ts"
import { recursiveAssign } from "./object.js"

Deno.test("no nested object", () => {
  const obj1 = { name: "test1", denug: true, num: 1, id: 1 }
  const obj2 = { name: "test2", denug: false, num: 2, ext: "more" }
  const obj3 = recursiveAssign({}, obj1, obj2)
  assertStrictEquals(obj3 === obj1, false)
  assertStrictEquals(obj3.id, obj1.id)
  assertStrictEquals(obj3.name, obj2.name)
  assertStrictEquals(obj3.denug, obj2.denug)
  assertStrictEquals(obj3.num, obj2.num)
  assertStrictEquals(obj3.ext, "more")

  const obj4 = recursiveAssign(obj1, obj2)
  assertStrictEquals(obj4, obj1)
})

Deno.test("nested object", () => {
  const obj1 = { nested: { name: "test1", denug: true, num: 1, id: 1 } }
  const obj2 = { nested: { name: "test2", denug: false, num: 2, ext: "more" } }
  const obj3 = recursiveAssign({}, obj1, obj2)
  assertStrictEquals(obj3 === obj1, false)
  assertStrictEquals(obj3.nested.id, obj1.nested.id)
  assertStrictEquals(obj3.nested.name, obj2.nested.name)
  assertStrictEquals(obj3.nested.denug, obj2.nested.denug)
  assertStrictEquals(obj3.nested.num, obj2.nested.num)
  assertStrictEquals(obj3.nested.ext, "more")
})


file.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.

/**
 * Fetch remote file with progress callback (start, received, end).
 *
 * Examples:
 * ```ts
 * import { Fetcher } from "https://deno.land/x/nextrj_utils@$VERSION/file.ts"
 *
 * const fetcher = new Fetcher(
 *   "https://www.example.com/test.pdf",
 *   {
 *     to: "example-test.pdf",
 *     on: {
 *       start(total: number) {
 *         console.log(`start: total=${total}`)
 *       },
 *       received(received: number, total: number) {
 *         console.log(`received: received=${received}, total=${total}`)
 *       },
 *       end(success: boolean) {
 *         console.log(`end: success=${success}`)
 *       },
 *     },
 *   },
 * )
 * await fetcher.fetch()
 * ```
 *
 * @module
 */

import { dirname, extension, joinPath, paseContentDisposition, writeAll } from "./deps.ts"
import { existsSync as pathExistsSync } from "./path.ts"
import { format as formatTemplate } from "./string.ts"

export type FetcherInit = {
  to?: string
  toIsNotTemplate?: boolean
  on?: {
    start?: (total: number, fileName: string, filePath: string) => void
    received?: (received: number, total: number) => void
    end?: (success: boolean) => void
  }
  fetchOptions?: RequestInit
}

/** A file fetcher with progress callback (start, received, end). */
export class Fetcher {
  #from: string
  #options?: FetcherInit
  constructor(from: string, options?: FetcherInit) {
    this.#from = from
    this.#options = options
  }
  /** Start fetch the file. */
  async fetch(): Promise<string> {
    const response = await fetch(this.#from, this.#options?.fetchOptions)
    if (response.ok) {
      // get file size from header
      let total = -1
      if (response.headers.has("Content-Length")) {
        total = parseInt(response.headers.get("Content-Length")!)
      }

      // get file name
      const fileName = getFileName(this.#from, response)
      // get the file full path name
      let filePath = this.#options?.to ? this.#options?.to : joinPath("temp", fileName)
      // format it: replace `\` to `/` to avoid template string erasing
      // in win os, path.join use `\` others use `/`
      if (!this.#options?.toIsNotTemplate) {
        filePath = formatTemplate(filePath.replace(/\\/g, "/"), { fileName: fileName })
      }

      // create the target dir if not exists
      const fileDir = dirname(filePath)
      if (!pathExistsSync(fileDir)) await Deno.mkdir(fileDir, { recursive: true })

      // pipe body to file
      const toFile: Deno.FsFile = await Deno.open(filePath, { write: true, create: true })
      const fetcher = this as Fetcher
      let received = 0 // sum chunk.byteLength
      const writableStream = new WritableStream({
        start(_controller) {
          // call start callback
          fetcher.#options?.on?.start?.call(this, total, fileName, filePath)
        },
        async write(chunk, controller) {
          try {
            await writeAll(toFile, chunk)
            received += chunk.byteLength
            // call received callback
            fetcher.#options?.on?.received?.call(this, received, total)
          } catch (e) {
            controller.error(e)
            toFile.close()
          }
        },
        close() {
          toFile.close()
          // call end callback for finish
          fetcher.#options?.on?.end?.call(this, true)
        },
        abort() {
          toFile.close()
          // call end callback for abort
          fetcher.#options?.on?.end?.call(this, false)
        },
      })
      await response.body?.pipeTo(writableStream)
      return filePath
    } else {
      response.body?.cancel()
      throw new Error(`${response.status} ${response.statusText}`)
    }
  }
}

/**
 * Extract the last path name from the url.
 *
 * throw {@link TypeError} if url is not a regular url.
 *
 * Examples: (decodeUri=true)
 * - "https://test.com/xxx" > "xxx"
 * - "https://test.com/xxx/" > "xxx"
 * - "https://test.com/" > "test.com"
 * - "file://path/to/xxx" > "xxx"
 * - "file://path/to/xxx/" > "xxx"
 * - "file://path/" > "path"
 * - "file://path/中文" > "中文"
 *
 * Examples: (decodeUri=false|undefined)
 * - "file://path/中文" > "中文"
 */
export function getLastPathName(url: string): string {
  return url.split("?")[0].replace(/\\/g, "/").split("/").filter(Boolean).pop() as string
}

/** Get filename param from content-disposition value. */
export function getFileNameFromContentDisposition(contentDisposition: string): string | undefined {
  try {
    const fileName = paseContentDisposition(contentDisposition)?.parameters?.filename
    // console.log(`ContentDisposition:filename=${filename}`)
    return fileName
  } catch {
    return undefined
  }
}

/**
 * Get the filename from response or request url.
 * 1. Get from content-disposition's filename* or filename param.
 * 2. Then get last path name from request url.
 * 3. Without extention, get from content-type.
 */
export function getFileName(requestUrl: string, response: Response): string {
  let fileName =
    // get from content-disposition header
    getFileNameFromContentDisposition(response.headers.get("content-disposition") || "") ||
    // get from request url
    getLastPathName(requestUrl)

  // without extention, get from content-type
  fileName = fileName.includes(".") ? fileName : `${fileName}.${extension(response.headers.get("content-type") || "")}`

  return decodeURIComponent(fileName)
}


path.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.

/** Extensions of "https://deno.land/std@$STD_VERSION/path/mod.ts".
 *
 * Codes in the examples check whether the path exists:
 *
 * ```ts
 * import { exists, existsSync } from "https://deno.land/x/nextrj_utils@$VERSION/path.ts"
 * const pathExists = await exists("./foo/bar")
 * const pathExistsSync = existsSync("./foo/bar")
 * console.log(pathExists);     // true or false
 * console.log(pathExistsSync); // true or false
 * ```
 *
 * @module
 */

/** Check whether the specified `path` exists.
 *
 * Requires `allow-read` permission.
 *
 * @tags allow-read
 * @category File System
 */
export async function exists(path: string | URL): Promise<boolean> {
  try {
    await Deno.lstat(path)
    return true
  } catch (err) {
    if (err instanceof Deno.errors.NotFound) return false
    else throw err
  }
}

/** Synchronously check whether the specified `path` exists.
 *
 * Requires `allow-read` permission.
 *
 * @tags allow-read
 * @category File System
 */
export function existsSync(path: string | URL): boolean {
  try {
    Deno.lstatSync(path)
    return true
  } catch (err) {
    if (err instanceof Deno.errors.NotFound) return false
    else throw err
  }
}


object.js
// ref from https://github.com/gmasmejean/recursiveAssign

function assign(source, key, value) {
  if (isPlainObject(value)) {
    if (!isPlainObject(source[key])) {
      source[key] = {}
    }
    mergeInObject(source[key], value)
  } else {
    source[key] = value
  }
}

function mergeInObject(target, source) {
  Object.keys(source).forEach((key) => {
    assign(target, key, source[key])
  })
}

function isPlainObject(o) {
  return o !== undefined && o.constructor !== undefined && o.constructor.prototype === Object.prototype
}

/** Recursive assign {@link sources} to {@link target} and return the {@link target}. */
export function recursiveAssign(target, ...sources) {
  if (typeof target === "object") {
    sources.forEach((source) => {
      if (isPlainObject(source)) {
        mergeInObject(target, source)
      }
    })
  }
  return target
}


duration.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.

/** Utilities for duration.
 *
 * Example for format seconds to `mm:ss` or `hh:mm:ss`:
 *
 * ```ts
 * import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
 * import { format } from "https://deno.land/x/nextrj_utils@$VERSION/duration.ts"
 *
 * assertStrictEquals(format(15), "00:15")
 * assertStrictEquals(format(2 * 60 + 15), "02:15")
 * assertStrictEquals(format(13 * 60 * 60 + 2 * 60 + 15), "13:02:15")
 * assertStrictEquals(format(24 * 60 * 60 + 2 * 60 + 15), "24:02:15")
 * ```
 *
 * @module
 */

const format2Hour = new Intl.DateTimeFormat("en-US", {
  hour12: false,
  hour: "2-digit",
  minute: "2-digit",
  second: "2-digit",
})
const format2Minute = new Intl.DateTimeFormat("en-US", {
  minute: "2-digit",
  second: "2-digit",
})

/**
 * Auto format {@link seconds} to `mm:ss` or `hh:mm:ss`.
 *
 * Example:
 * ```ts
 * format(2 * 60 + 15) // "02:15"
 * format(1 * 60 * 60 + 2 * 60 + 15) // "01:02:15"
 * format(13 * 60 * 60 + 2 * 60 + 15) // "13:02:15"
 * ```
 */
export function format(seconds: number): string {
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  return (hours >= 1 ? format2Hour : format2Minute)
    .format(new Date(0, 0, 0, hours, minutes % 60, seconds % 60))
}


terminal_progress_bar.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.

/** Terminal progress bar */

export type Options = {
  value?: number
  total?: number
  title?: string
  template?: string
}

const encode = (content: string) => new TextEncoder().encode(content)

export const TerminalProgressBar = {
  title: undefined,
  value: 0,
  total: 100,
  get completed() {
    return this.value >= this.total
  },
  lastContent: "",
  percentPrecision: 2,
  template: "${bar} ${duration} ${percent} ${title} ${value}/${total}",
  init(options?: Options | number) {
    if (typeof options === "number") {
      this.value = Math.max(0, options)
      this.total = 100
    } else if (typeof options === "undefined") {
      this.value = 0
      this.total = 100
    } else throw Error("Not implements")
  },
  to(options: Options | number) {
    if (typeof options === "number") {
      this.value = Math.min(this.total, options)
    } else throw Error("Not implements")

    // generate the content
    const content = `${this.value}/${this.total}`

    // only render when content changed
    if (this.lastContent !== content) {
      this.lastContent = content
      // write text to console
      // 1. `\r` - move cursor to the left egde
      // 2. `\x1b[?25l` - hide the cursor
      // 3. `\x1b[0K` - clear from the content tail to the right edge
      Deno.stdout.writeSync(encode(`\r${content}\x1b[?25l\x1b[0K`))
    }
  },
  clear(keepState = true) {
    if (keepState) {
      // break line and show cursor, not clear last content
      // 1. `\n` - break line
      // 2. `\r` - move cursor to left edge
      // 3. `\x1b[?25h` - show cursor (h-height bit, l-low bit)
      Deno.stdout.writeSync(encode(`\n\r\x1b[?25h`))
    } else {
      // clear content and show cursor
      // 1. `\x1b[2K` - clear content from left edge to right edge
      // 2. `\r` - move cursor to left edge
      // 3. `\x1b[?25h` - show cursor (h-height bit, l-low bit)
      Deno.stdout.writeSync(encode("\x1b[2K\r\x1b[?25h"))
    }
  },
}


string.ts
// Copyright 2023 the NextRJ organization. All rights reserved. MIT license.

/**
 * Utilities for string.
 *
 * Example for format string template:
 *
 * ```ts
 * import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
 * import { format } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"
 *
 * assertStrictEquals(format("${a} / ${b}", { a: 50, b: 100 }), "50 / 100")
 *
 * const f = (v: number): number => ++v
 * assertStrictEquals(format("${v}-${f(1)}", { v: 1, f }), "1-2")
 * ```
 *
 * Example for truncate filename:
 *
 * ```ts
 * import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
 * import { truncateFilename } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"
 *
 * truncateFilename("123456789.zip", 10) // "12...9.zip"
 * // 🦄 is 2 column width
 * truncateFilename("1234567🦄.zip", 10) // "1...🦄.zip"
 * ```
 *
 * Example for truncate string:
 *
 * ```ts
 * import { assertStrictEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
 * import { truncate } from "https://deno.land/x/nextrj_utils@$VERSION/string.ts"
 *
 * const maxLen = 10
 * console.log(truncate("=".repeat(1000), maxLen))
 * console.log(truncate("±".repeat(1000), maxLen))
 * console.log(truncate("★".repeat(1000), maxLen))
 * console.log(truncate("中".repeat(1000), maxLen))
 * console.log(truncate("🦄".repeat(1000), maxLen))
 * console.log(truncate("🦄".repeat(1000), maxLen))
 * console.log(truncate("=中±🦄±中=".repeat(1000), maxLen))
 *
 * // they should all have the very similar visable width in the terminal:
 * ------- output -------
 * ==========
 * ±±±±±±±±±±
 * ★★★★★★★★★★
 * 中中中中中
 * 🦄🦄🦄🦄🦄
 * 🦄🦄🦄🦄🦄
 * =中±🦄±中=
 * ----- output end -----
 * ```
 *
 * @module
 */

import { extname, stringWidth } from "./deps.ts"

// only for UTF8 https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder
const textEncoder = new TextEncoder()

/**
 * Count the number of words in a string.
 * One word (even multi-byte string) has one length.
 *
 * It is not the same with code-unit-length of `String.length` for multi-byte string.
 * The count is base on the code-point-length.
 * See [<How big is a JavaScript string?>](https://blog.bitsrc.io/how-big-is-a-string-ef2af3d222e6).
 *
 * - wordCount("a中🦄") // 3
 * - comparison with code-unit-length:
 *     - "a中🦄".length // 4
 *     - "a".length // 1
 *     - "中".length // 1
 *     - "🦄".length // 2
 */
export function wordCount(source: string): number {
  // same with `Array.from(source).length`
  return [...source].length
}

/**
 * Count the visable width of a string. Similar with terminal columns.
 *
 * - ASCII character has 1 with.
 * - Chiness character and Emoji have 2 width.
 *
 * - columnCount("a") // 1
 * - columnCount("±") // 1
 * - columnCount("中") // 2
 * - columnCount("🦄") // 2
 */
export function columnCount(source: string): number {
  return stringWidth(source)
}

/**
 * Get the specified `source` string's byte-length.
 *
 * - byteLength("a") // 1
 * - byteLength("±") // 2
 * - byteLength("★") // 3
 * - byteLength("中") // 3
 * - byteLength("🦄") // 4
 */
export function byteLength(source: string): number {
  return textEncoder.encode(source).length
}

export enum MaxLenType {
  MaxColumnCount,
  MaxByteLength,
  MaxWordCount,
}

/**
 * Truncate the specified `source` string to a max-length string.
 * Defaults truncate by max-column-count for better match visable with.
 */
export function truncate(
  source: string,
  maxLen: number,
  maxLenType: MaxLenType = MaxLenType.MaxColumnCount,
): string {
  if (source.length === 0) return source
  if (maxLen < 0) throw new Error(`Argument 'maxLen' should greater than 0.`)

  // split source to code array by code-point.
  // don't use `source.split("")` because it use code-unit-length.
  // see https://blog.bitsrc.io/how-big-is-a-string-ef2af3d222e6
  const codes = [...source]
  const codesLen = codes.length

  // truncate it
  if (maxLenType === MaxLenType.MaxColumnCount) { // truncate by max-column-count
    if (columnCount(source) > maxLen) {
      const ss = []
      let count = 0
      for (const code of codes) {
        // console.log("code=" + code)
        count += columnCount(code)
        if (count <= maxLen) ss.push(code)
        else break
      }
      return ss.join("")
    } else return source
  } else if (maxLenType === MaxLenType.MaxByteLength) { // truncate by max-byte-length
    if (byteLength(source) > maxLen) {
      const ss = []
      let count = 0
      for (const code of codes) {
        count += byteLength(code)
        if (count <= maxLen) ss.push(code)
        else break
      }
      return ss.join("")
    } else return source
  } else if (maxLenType === MaxLenType.MaxWordCount) { // truncate by max-code-point-length
    if (codesLen > maxLen) return codes.slice(0, maxLen).join("")
    else return source
  } else throw new Error(`Unsupport maxLenType of '${maxLenType}'`)
}

/**
 * Truncate the filename to a max-length string with its extension.
 *
 * Rule:
 * 1. Keep the extension name.
 * 2. Truncate the middle string near by the extension.
 * 3. Use ellipsis symbol `...` to replace the truncated part.
 *
 * Examples:
 *
 * ```
 * truncateFilename("123456789.zip", 10) // "12...9.zip"
 * truncateFilename("12345678🦄.zip", 10) // "1...🦄.zip"
 * ```
 */
export function truncateFilename(
  fileName: string,
  maxLen: number,
): string {
  if (maxLen >= columnCount(fileName)) return fileName
  // a/b/123456789🦄.zip:
  // 1. ext=".zip"
  // 2. name="123456789"
  // 3. lastKeepPart="🦄.zip"
  // 4. toTruncatePart="123456789"
  const ext = extname(fileName)
  if (ext === "") return truncate(fileName, maxLen, MaxLenType.MaxColumnCount)
  const name = fileName.substring(0, fileName.length - ext.length)
  const codes = [...name]
  const lastKeepPart = codes[codes.length - 1] + ext
  const toTruncatePart = codes.slice(0, codes.length - 1).join("")

  const maxLen2 = maxLen - columnCount(lastKeepPart)
  // console.log(`ext=${ext}, lastKeepPart=${lastKeepPart}, toTruncatePart=${toTruncatePart}, maxLen2=${maxLen2}`)
  if (maxLen2 <= 3 + columnCount(codes[0])) return codes[0] + "..." + lastKeepPart
  else return truncate(toTruncatePart, maxLen2 - 3, MaxLenType.MaxColumnCount) + "..." + lastKeepPart
}

/**
 * Format the {@link template} with the specific {@link params}.
 *
 * Example:
 * ```ts
 * format("${a}/${b}", {a: 50, b: 100}) // "50/100"
 * ```
 */
export function format(template: string, params: Record<string, unknown>): string {
  const fn = new Function(...Object.keys(params), "return `" + template + "`;")
  return fn(...Object.values(params))
}
